recipe/meta.yaml

```
@@ -6,12 +6,12 @@ package:
   version: {{ version }}
 
 source:
-  url: https://github.com/PolusAI/{{ name }}/archive/refs/tags/{{ version }}.tar.gz
-  sha256: f3041bb73ccb6390df5c2cf6d8ac7d4f5d91b75a3f31dd7a0664763d964c5138
+  url: https://github.com/sameeul/nyxus/archive/refs/heads/conda_gpu_v3.zip
+  sha256: 2dbd89c9f501410631413ed88fbc91c1ebfdaf4e439a28420501fb9745a70695  
   patches:
     - 0000-win_2022.patch
 build:
-  number: 0
+  number: 1
   skip: true  # [py<36]
```

This is how we would skip a CUDA version (like 11.2)

Would also want to re-render after making this change to update the CI matrix accordingly

```suggestion
  skip: true  # [py<36 or cuda_compiler_version == "11.2"]
```

---

Also if we want to only test CUDA 12 for a bit, would try something like this

Again would re-render after making a change like this to update CI

```suggestion
  skip: true  # [py<36 or cuda_compiler_version != "12.0"]
```

---

Thanks. Also, do I need anything else to get the CUDA 12 compiler detected. Please see the following output from CUDA 11.2 and CUDA 12 run. CMake fails to find the compiler in CUDA 12 case. Do I need to install any other package?

11.2
```
2024-02-08T18:04:44.5335279Z   -- Looking for a CUDA compiler
2024-02-08T18:04:46.7857213Z   -- Looking for a CUDA compiler - /home/conda/feedstock_root/build_artifacts/nyxus_1707415201231/_build_env/bin/nvcc
2024-02-08T18:04:46.8232304Z   -- Found CUDAToolkit: /usr/local/cuda/targets/x86_64-linux/include (found version "11.2.152")
```

and 12.0
```
2024-02-08T18:15:45.8472644Z   -- Looking for a CUDA compiler
2024-02-08T18:15:47.1839760Z   -- Looking for a CUDA compiler - NOTFOUND
2024-02-08T18:15:47.2097249Z   -- Found CUDAToolkit: /home/conda/feedstock_root/build_artifacts/nyxus_1707415928413/_build_env/targets/x86_64-linux/include (found version "12.0.76")
```

---

Could you please share a bit more about the changes that occurred in the build system?

---

Here are the changes we made in `CMakeLists.txt`. 

Previously,  by default, `USEGPU` was set to `OFF`. CMake would look for CUDA compiler (by using `check_language(CUDA`) and CUDA toolkit (by using `find_package(CUDAToolkit)`). If both were true, it would set `USEGPU` to `ON`, which will ensure all the `*cu` files will be compiled. If the user passed `-DUSEGPU=ON`  in `CMAKE_ARGS` and the aforementioned checks failed, it would just emit a warning saying it could not find CUDA , set `USEGPU` to `OFF` and move on.

We have modified the behavior the following way now: `USEGPU` is set to `OFF` be default. Now, CMake will look for CUDA only if the user sets `USEGPU` to `ON`. It will check for CUDA compiler and CUDATooklit (same way as mentioned earlier) and if any of these check fails, it will issue a `FATAL` message saying CUDA could not be found.

On the feedstock side, we modified `build.sh` and `bld.bat` to look for `nvcc` command (in this PR as test) and based on that pass `-DUSEGPU=ON` in the `CMAKE_ARGS`. So, for CUDA 11.2/11.8/12.0, the `build.sh` correctly identifies the presence of `nvcc` and adds `-DUSEGPU=ON` to `CMAKE_ARGS`. However, we see that for CUDA 12 builds, the `check_language(CUDA)` call fails to find the `nvcc` compiler. However, in 11.2 and 11.8, it works. So, I am trying to understand whether I need to add any other package/meta-package in the `host` section to make sure CMake finds CUDA.

It is also working as expected for Windows builds. 

Also, when we first added CUDA 12 support, I verified from the logs that things were compiled correctly. However, I did not do that in our last release (~Dec 23). So, I can not say when this regression occurred, since the previous behavior was to emit a warning and keep moving. 

Hope this answers your question.

---

looks like I had to add `- cuda-version {{ cuda_compiler_version }} # [cuda_compiler_version != "None"]` to the host section to make CMake find the CUDA compiler.

---

Ah ok. That makes more sense

Yeah `cuda-version` needs to be constrained in `requirements/host` to align it with `requirements/build` so they have the same CUDA compiler version

We have had some discussion about adding functionality to conda & conda-build to automate this